File: backend/src/main/java/com/example/chat/controller/AuthController.java
package com.example.chat.controller;

import com.example.chat.service.AuthService;
import com.example.chat.web.dto.AuthDtos;
import com.example.chat.entity.User;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/auth")
public class AuthController {

    @Autowired
    private AuthService authService;

    @Autowired
    private AuthenticationManager authenticationManager;

    @PostMapping("/register")
    public ResponseEntity<?> register(@RequestBody AuthDtos.RegisterRequest registerRequest) {
        User registeredUser = authService.register(registerRequest);
        AuthDtos.UserDto userDto = new AuthDtos.UserDto(registeredUser.getId(), registeredUser.getUsername(), registeredUser.getEmail());
        return ResponseEntity.ok(userDto);
    }

    @PostMapping("/login")
    public ResponseEntity<?> login(@RequestBody AuthDtos.LoginRequest loginRequest) {
        User user = authService.login(loginRequest);
        if (user == null) {
            return ResponseEntity.status(401).body("Invalid email or password");
        }
        String token = authService.generateTokenFromUser(user);
        AuthDtos.UserDto userDto = new AuthDtos.UserDto(user.getId(), user.getUsername(), user.getEmail());
        return ResponseEntity.ok(new AuthDtos.LoginResponse(token, userDto));
    }
}

File: backend/src/controllers/authController.ts
import { Request, Response } from 'express';
import { authService } from '../services/authService';
import { RegisterRequest, LoginRequest } from '../models/User';

export class AuthController {
  public async register(req: Request, res: Response): Promise<void> {
    try {
      const { username, email, password }: RegisterRequest = req.body;

      if (!username || !email || !password) {
        res.status(400).json({ error: 'Username, email, and password are required' });
        return;
      }

      const userDto = await authService.register({ username, email, password });
      res.status(200).json(userDto);
    } catch (error: any) {
      res.status(400).json({ error: error.message || 'Registration failed' });
    }
  }

  public async login(req: Request, res: Response): Promise<void> {
    try {
      const { email, password }: LoginRequest = req.body;

      if (!email || !password) {
        res.status(400).json({ error: 'Email and password are required' });
        return;
      }

      const loginResponse = await authService.login({ email, password });

      if (!loginResponse) {
        res.status(401).send('Invalid email or password');
        return;
      }

      res.status(200).json(loginResponse);
    } catch (error: any) {
      res.status(500).json({ error: error.message || 'Login failed' });
    }
  }
}

export const authController = new AuthController();
